core/domain/stats_services.py

```diff
@@ -212,6 +213,165 @@ def handle_stats_creation_for_new_exp_version(
+        keyname = stats_models.ISSUE_TYPE_KEYNAME_MAPPING[exp_issue.issue_type]
+        # If keyname is 'state_names', which is a list, we add an instance of
+        # this issue to all associated states in the mapping.
+        if keyname == 'state_names':
+            state_names = exp_issue.issue_customization_args[keyname]['value']
+            for state_name in state_names:
+                exp_issues_by_state_name[state_name].append(exp_issue)
+        else:
+            state_name = exp_issue.issue_customization_args[keyname]['value']
+            exp_issues_by_state_name[state_name].append(exp_issue)
+
+    exp_versions_diff = exp_domain.ExplorationVersionsDiff(change_list)
+
+    # Handling deletions.
+    for state_name in exp_versions_diff.deleted_state_names:
+        for exp_issue in exp_issues_by_state_name[state_name]:
+            idx = exp_issues.unresolved_issues.index(exp_issue)
+            exp_issues.unresolved_issues[idx].is_valid = False
+        save_exp_issues_model_transactional(exp_issues)
+
+    # Handling renames.
+    for old_state_name in exp_versions_diff.old_to_new_state_names:
+        new_state_name = exp_versions_diff.old_to_new_state_names[
+            old_state_name]
+        playthrough_ids = []
+        for exp_issue in exp_issues_by_state_name[old_state_name]:
+            idx = exp_issues.unresolved_issues.index(exp_issue)
+            # Modify occurences in exploration issues.
+            if stats_models.ISSUE_TYPE_KEYNAME_MAPPING[
+                    exp_issue.issue_type] == 'state_names':
+                state_names = exp_issue.issue_customization_args['state_names'][
+                    'value']
+                exp_issue.issue_customization_args['state_names']['value'] = [
+                    new_state_name
+                    if state_name == old_state_name else state_name
+                    for state_name in state_names]
+            else:
+                exp_issue.issue_customization_args['state_name']['value'] = (
+                    new_state_name)
+            playthrough_ids.extend(exp_issue.playthrough_ids)
+
+        # Modify occurences in playthroughs.
+        playthroughs = get_playthroughs_multi(playthrough_ids)
+        for idx, playthrough in enumerate(playthroughs):
+            if stats_models.ISSUE_TYPE_KEYNAME_MAPPING[
+                    playthrough.issue_type] == 'state_names':
+                state_names = playthrough.issue_customization_args[
+                    'state_names']['value']
+                playthrough.issue_customization_args['state_names']['value'] = [
+                    new_state_name
+                    if state_name == old_state_name else state_name
+                    for state_name in state_names]
+            else:
+                playthrough.issue_customization_args['state_name']['value'] = (
+                    new_state_name)
+            for idx1, action in enumerate(playthrough.playthrough_actions):
```

Comment:
Just wondering, btw -- is there a reason to call this playthrough.playthrough_actions instead of playthrough.actions? Looks a little odd.

Reply:
@seanlip Hmm, just actions might look nice too. I'll change this in that one follow up PR after Milestone 1 where I have to change a bunch of stuff. I'll have it all written on my Devlogs.